### PR TITLE
S3 Watcher

### PR DESCRIPTION
**Pull request by kenoir**
Intended to replace/supplement ftp-watcher by allowing uploads to s3 bucket.

Depends: https://github.com/guardian/grid-infra/pull/103/files

### CODE REVIEW

File: deploy_s3watcher.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#/bin/bash
```
Comment (theefer) [inline deploy_s3watcher.sh]: Not urgent, but would be lovely to have a standardised way to push a build to S3. Would be useful to deploy kahuna as static files, or other things like the wat pages and grid-inspector. I think there may be an sbt plugin that does this but obviously a lot of these things are *not* sbt projects, so would have to be more generic.

Just a thought.

File: s3watcher/package.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    "async": "~1.4.2",
+    "fs": "0.0.2",
+    "request": "~2.61.0"
+  },
+  "devDependencies": {},
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "BSD-2-Clause"
```
Comment (theefer) [inline s3watcher/package.json]: Our default is typically Apache-2

**Comment by theefer** [comment]
Exciting :-)

File: deploy_s3watcher.sh (hunk 1 above)
Comment (kenoir) [inline deploy_s3watcher.sh]: I wonder if @sihil has any thoughts on deploying lambda functions & deploying static sites via riffraff.

File: s3watcher/package.json (hunk 2 above)
Comment (kenoir) [inline s3watcher/package.json]: This is the default generated by npm. Will fix.

File: s3watcher/S3Watcher.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+exports.handler = function(event, context) {
+    var srcBucket = event.Records[0].s3.bucket.name;
+    var srcKey = decodeURIComponent(event.Records[0].s3.object.key.replace(/\+/g, " "));
+    var splitSrcKey = srcKey.split("/");
+    var simpleKey = splitSrcKey[(splitSrcKey.length - 1)];
+    var filename = '/tmp/' + simpleKey;
+
+    // TODO: Constants we want set from elsewhere!
+    var baseUrl = "nope";
```
Comment (kenoir) [inline s3watcher/S3Watcher.js]: @gidsg do we have an established pattern for injecting secrets into Lambda functions? I was thinking perhaps pull and parse a properties file from S3.

File: s3watcher/S3Watcher.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+var Rx = require('rx');
+
+var Lambda   = require('./lib/Lambda');
+var Transfer = require('./lib/Transfer');
+
+
+process.env['NODE_TLS_REJECT_UNAUTHORIZED'] = '0';
```
Comment (theefer) [inline s3watcher/S3Watcher.js]: Does this disable TLS verification? Do we want to have that in PROD?

Comment (kenoir) [inline s3watcher/S3Watcher.js]: Yes & no.

File: s3watcher/lib/Upload.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            body: data,
+            forever: true
+        };
+
+        const uploadRequest = request.post(options);
+
+        return Rx.Observable.create(function(observer){
+           uploadRequest.on("response", function(res){
+               observer.onNext(res);
+           }.bind(observer));
```
Comment (theefer) [inline s3watcher/lib/Upload.js]: Why is the extra `function` required here?

File: s3watcher/lib/Upload.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    buildUpload: function(config, s3Event) {
+         const headers = {
+            'Content-Length': s3Event.size,
+            'Content-Type': 'application/octet-stream',
+            'X-Gu-Media-Key': config.apiKey
+        };
+
+        return  {
+            key: config.apiKey,
+            url: config.baseUrl + "/images?filename=" + s3Event.filename,
```
Comment (theefer) [inline s3watcher/lib/Upload.js]: Could the filename be passed along as a `params` to the `request` call, to ensure proper encoding of it?

We will need to send the `uploadedBy` information too (folder?).

File: s3watcher/lib/Upload.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            size: s3Event.size,
+            headers: headers
+        };
+    },
+    postData: function(upload, data) {
+        const options = {
+            url: upload.url,
+            headers: upload.headers,
+            body: data,
+            forever: true
```
Comment (theefer) [inline s3watcher/lib/Upload.js]: Worth documenting why this is needed?

File: s3watcher/S3Watcher.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+exports.handler = function(event, context) {
+
+    const lambda   = Lambda.init(event, context);
+    const transfer = Transfer.init(lambda.s3Event);
+
+    const operation = lambda.config.flatMap(transfer.operation).share();
+
+    operation.subscribeOnError(lambda.fail);
```
Comment (theefer) [inline s3watcher/S3Watcher.js]: Is this necessary? Should be caught by line 23 and 29?

File: s3watcher/lib/Transfer.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+            console.log("Deleting from ingest bucket.");
+            return S3Helper.deleteS3Object({
+                Bucket: s3Event.bucket,
+                Key: s3Event.key
+            });
+        };
+
+        const fail = function(result) {
+            console.log("Upload failed with status code: " + result.response.statusCode);
+            console.log("Copying to fail bucket.");
```
Comment (theefer) [inline s3watcher/lib/Transfer.js]: Note that in the FTP watcher we tried three times.

File: s3watcher/lib/Lambda.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+            const srcSize = e.object.size;
+            const srcFilename = filenameFromKey(srcKey);
+
+            return {
+                bucket: srcBucket,
+                key: srcKey,
+                filename: srcFilename,
+                size: srcSize
+            };
+        })();
```
Comment (theefer) [inline s3watcher/lib/Lambda.js]: Bit odd to have an IIFE, could be split into an actual function?

File: s3watcher/lib/Transfer.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
+var S3Helper     = require('./S3Helper');
+var Upload       = require('./Upload');
+
+
+module.exports = {
+    init: function(s3Event){
+        const operation = function(config) {
+            console.log("Downloading from ingest bucket.");
+            const upload = Upload.buildUpload(config, s3Event);
```
Comment (theefer) [inline s3watcher/lib/Transfer.js]: Redundant with line 16.

File: s3watcher/lib/Lambda.js (hunk 10 above)
Comment (kenoir) [inline s3watcher/lib/Lambda.js]: Do you mean it's a bit IIFEy? *rimshot*

File: s3watcher/S3Watcher.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
+var Rx = require('rx');
```
Comment (jamesgorrie) [inline s3watcher/S3Watcher.js]: These could be `const`s.

**Comment by akash1810** [comment]
Not part of this PR, but would we able to add a filter to the objects we get from s3 to fix https://github.com/guardian/grid/issues/573?

**Comment by theefer** [comment]
I don't think #573 will be an issue when we use this, as it's specifically a bug in the FTPWatcher.

File: s3watcher/lib/Transfer.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+            console.log("Downloading from ingest bucket.");
+
+            const upload = Upload.buildUpload(config, s3Event);
+
+            return S3Helper.getS3Object({
+                Bucket: s3Event.bucket,
+                Key: s3Event.key
+            }).flatMap(function(data){
+                console.log("Uploading to image-loader.");
+
```
Comment (theefer) [inline s3watcher/lib/Transfer.js]: Could put in a TODO hint to try and use the stream API for better performance (with the caveat that it's a big PITA)?

**Comment by theefer** [comment]
Great stuff! :+1: 

Let's prove in TEST and talk to Adrian about getting this properly setup!

**Pull request merged**